packages/go_router/lib/src/route.dart

```diff
@@ -1255,8 +1291,39 @@ class StatefulNavigationShellState extends State<StatefulNavigationShell>
     final bool locationChanged =
         previousBranchLocation != currentBranchLocation;
     if (locationChanged || !hasExistingNavigator) {
-      _branchNavigators[branch.navigatorKey] = shellRouteContext
-          .navigatorBuilder(branch.observers, branch.restorationScopeId);
+      _branchNavigators[branch.navigatorKey] =
```

Comment:
just realized somewhere in this method should probably remove entries in _branchNavigators the are no longer in the StatefulShellRoute?

Reply:
You're right, that is definitely necessary to do now. Will have a look at it.

Reply:
@chuntai, I added a fix for removing old entries, but when attempting to write a test case, I ran into issues with duplicate GlobalKeys (_shellStateKey) when updating the dynamic RoutingConfig. Or - I should be clear - at first the test case worked, but not because of the code I added, but instead because the entire StatefulShellRoute was replaced/reloaded. But when I added the possibility to set the `GlobalKey<StatefulNavigationShellState> _shellStateKey` to maintain state, I started to get this error.  

When trying to troubleshoot it, I started to wonder whether dynamic RoutingConfig full supports ShellRoutes, at least with state. Added some test cases to test this, and it seems simple StatefulWidgets in a simple routing scenario works, but when adding a (regular) ShellRoute with a stateful shell I also get and issue with duplicate GlobalKeys. 

Not sure if I just missed something, but I anyway added a gist with the test cases here: https://gist.github.com/tolo/922fd838cdea30b17121533b0012eda4

Reply:
can you file an issue for it? looks like a bug

Reply:
is the fix in this pr already?

Reply:
It is now. 😊 Note though that I disabled two tests that fail due to the issue I mentioned above.